src/overlays/actors/ovl_Efc_Erupc/z_efc_erupc.c

```diff
@@ -26,23 +31,230 @@ const ActorInit Efc_Erupc_InitVars = {
+            func_80093D84(globalCtx->state.gfxCtx);
+            gSPDisplayList(POLY_XLU_DISP++, D_06002760);
+            gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, particles->color.r, particles->color.g, particles->color.b,
+                            particles->alpha);
+            gDPSetEnvColor(POLY_XLU_DISP++, 150, 0, 0, 0);
+            gDPPipeSync(POLY_XLU_DISP++);
+            Matrix_Translate(particles->pos.x, particles->pos.y, particles->pos.z, MTXMODE_NEW);
+            func_800D1FD4(&globalCtx->mf_11DA0);
+            Matrix_Scale(particles->scale, particles->scale, 1.0f, MTXMODE_APPLY);
+            gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gfxCtx, "../z_efc_erupc.c", 393),
+                      G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+            gSPDisplayList(POLY_XLU_DISP++, D_060027D8);
+        }
+    }
+    CLOSE_DISPS(gfxCtx, "../z_efc_erupc.c", 399);
+}
+
+Color_RGB8 D_8099D770[] = {
+    { 255, 128, 0 },
+    { 255, 0, 0 },
+    { 255, 255, 0 },
+    { 255, 0, 0 },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Efc_Erupc/EfcErupc_Update.s")
+void EfcErupc_UpdateParticles(EfcErupc* this, GlobalContext* globalCtx) {
+    s16 i;
+    s16 index;
+    Color_RGB8 particleColors[] = {
+        { 255, 128, 0 },
+        { 255, 0, 0 },
+        { 0, 0, 0 },
+        { 100, 0, 0 },
+    };
+    Color_RGB8* color;
+    EfcErupcParticles* cur = this->particles;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Efc_Erupc/EfcErupc_Draw.s")
+    for (i = 0; i < EFC_ERUPC_NUM_PARTICLES; i++, cur++) {
+        if (cur->isActive) {
+            cur->pos.x += cur->vel.x;
+            cur->pos.y += cur->vel.y;
+            cur->pos.z += cur->vel.z;
+            cur->vel.x += cur->accel.x;
+            cur->vel.y += cur->accel.y;
+            cur->vel.z += cur->accel.z;
+            cur->animTimer += 1;
+            index = cur->animTimer % 4;
+            color = &particleColors[index];
+            cur->color.r = color->r;
+            cur->color.g = color->g;
+            cur->color.b = color->b;
+            cur->alpha -= 20;
+            if (cur->alpha <= 0) {
+                cur->alpha = 0;
+                cur->isActive = false;
+            }
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Efc_Erupc/func_8099D334.s")
+void EfcErupc_AddParticle(EfcErupcParticles* particles, Vec3f* pos, Vec3f* vel, Vec3f* accel, f32 scaleFactor) {
+    s16 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Efc_Erupc/func_8099D52C.s")
+    for (i = 0; i < EFC_ERUPC_NUM_PARTICLES; i++, particles++) {
+        if (!particles->isActive) {
+            particles->isActive = true;
+            particles->pos = *pos;
+            particles->vel = *vel;
+            particles->accel = *accel;
+            particles->scale = scaleFactor / 1000.0f;
+            particles->alpha = 255;
+            particles->animTimer = (s8)(s32)(s8)(s32)Rand_ZeroFloat(10);
```

Comment:
This seems really fake but until someone has a better idea I guess its OK.  Maybe leave a note though.